packages/artifact/src/upload-artifact-http-client.ts

```diff
@@ -0,0 +1,299 @@
+    Name: artifactName
+  }
+  const data: string = JSON.stringify(parameters, null, 2)
+  const rawResponse: HttpClientResponse = await client.post(
+    getArtifactUrl(),
+    data,
+    requestOptions
+  )
+
+  const body: string = await rawResponse.readBody()
+  const response: CreateArtifactResponse = JSON.parse(body)
+  console.log(response)
+
+  if (rawResponse.message.statusCode === 201 && response) {
+    return response
+  } else {
+    throw new Error(
+      'Non 201 status code when creating file container for new artifact'
+    )
+  }
+}
+
+/**
+ * Step 2 of 3 when uploading an artifact. Concurrently upload all of the files in chunks
+ * @param {string} uploadUrl Base Url for the artifact that was created
+ * @param {SearchResult[]} filesToUpload A list of information about the files being uploaded
+ * @returns The size of all the files uploaded in bytes
+ */
+export async function uploadArtifactToFileContainer(
+  uploadUrl: string,
+  filesToUpload: SearchResult[]
+): Promise<number> {
+  const token = process.env['ACTIONS_RUNTIME_TOKEN'] || ''
+  const bearerCredentialHandler = new BearerCredentialHandler(token)
+  const client: HttpClient = new HttpClient('actions/artifact', [
+    bearerCredentialHandler
+  ])
+
+  const FILE_CONCURRENCY =
+    parseEnvNumber('ARTIFACT_FILE_UPLOAD_CONCURRENCY') ||
+    defaultFileUploadConcurrency
+  const CHUNK_CONCURRENCY =
+    parseEnvNumber('ARTIFACT_CHUNK_UPLOAD_CONCURRENCY') ||
+    defaultChunkUploadConcurrency
+  const MAX_CHUNK_SIZE =
+    parseEnvNumber('ARTIFACT_UPLOAD_CHUNK_SIZE') || 4 * 1024 * 1024 // 4 MB Chunks
+  debug(
+    `File Concurrency: ${FILE_CONCURRENCY}, Chunk Concurrency: ${CHUNK_CONCURRENCY} and Chunk Size: ${MAX_CHUNK_SIZE}`
+  )
+
+  const parameters: UploadFileParameters[] = []
+
+  // Prepare the necessary parameters to upload all the files
+  for (const file of filesToUpload) {
+    const resourceUrl = new URL(uploadUrl)
+    resourceUrl.searchParams.append(
+      'scope',
+      '00000000-0000-0000-0000-000000000000'
+    )
+    resourceUrl.searchParams.append('itemPath', file.uploadFilePath)
+    parameters.push({
+      file: file.absoluteFilePath,
+      resourceUrl: resourceUrl.toString(),
+      restClient: client,
+      concurrency: CHUNK_CONCURRENCY,
+      maxChunkSize: MAX_CHUNK_SIZE
+    })
+  }
+
+  const parallelUploads = [...new Array(FILE_CONCURRENCY).keys()]
+  const fileSizes: number[] = []
+  let uploadedFiles = 0
+
+  // Only allow a certain amount of files to be uploaded at once, this is done to reduce errors if
+  // trying to upload everything at once
+  await Promise.all(
+    parallelUploads.map(async () => {
+      while (uploadedFiles < filesToUpload.length) {
+        const currentFileParameters = parameters[uploadedFiles]
+        uploadedFiles += 1
+        fileSizes.push(await uploadFileAsync(currentFileParameters))
+      }
+    })
+  )
+
+  let sum = 0
+  for (const fileSize of fileSizes) {
+    sum += fileSize
+  }
```

Comment:
minor: Can use `Array.Reduce` to calculate this sum:
```ts
const sum = fileSizes.reduce((acc, val) => acc + val);
```

Reply:
![image](https://user-images.githubusercontent.com/16109154/72391824-52fa1100-36fc-11ea-9218-f9b55fe589e6.png)


Fancy 😀 Didn't know about this `reduce` function

Reply:
Ended up doing a massive overhaul around this part of the code so I no longer need to sum up arrays. Nice tip though!